JDK-8266503: [UL] Make Decorations safely copy-able and reduce their size

tstuefe opened the pull request:
This patch reduces the size of UL `LogDecorations` by about 85% (from 368 -> 56 bytes on Linux x64). This matters in the context of asynchronous logging in UL where we plan to keep a buffer containing log messages, including decorations, for asynchronous printing.

As a side effect, it makes the LogDecorations object safe to copy with trivial assignment constructors and operators (which it had not been before).

As another side effect, the 256-char-for-all-decorators limit has been removed with this patch.

What the patch does:

In LogDecorations, we resolve the values of the given decorators ("uptime", "tid" etc) and print them in human-readable format. Before this patch, the class LogDecorations stored the printed decorators in an internal (limited, fixed-sized) buffer. This is inefficient since this takes much more memory than storing the binary data before printing them.

So this patch separates the decorator value resolving from the printing. It stores the resolved values in binary form and only prints them when needed. Since a decorations object is only printed once this is fine. No need to cache the formatted text.

*Please Note that this patch includes the fix for JDK-8266536: "Provide a variant of os::iso8601_time which works with arbitrary timestamps" which is in a separate PR (https://github.com/openjdk/jdk/pull/3869) - so please ignore all the iso6801 stuff. It also contains a fix for the broken harfbuzz build.*

Testing:
- gtests, manually
- jtreg runtime/logging, manually
- SAP nightlies ran fine, but since then the tests changed, will run them again tonight

<!--
Replace this text with a description of your pull request (also remove the surrounding HTML comment markers).
If in doubt, feel free to delete everything in this edit box first, the bot will restore the progress section as needed.
-->

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266503](https://bugs.openjdk.java.net/browse/JDK-8266503): [UL] Make Decorations safely copy-able and reduce their size


### Reviewers
 * [Volker Simonis](https://openjdk.java.net/census#simonis) (@simonis - **Reviewer**)
 * [Yasumasa Suenaga](https://openjdk.java.net/census#ysuenaga) (@YaSuenag - **Reviewer**) ⚠️ Review applies to 3a35755702f93e5875458590d79419e6d02a5524


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3874/head:pull/3874` \
`$ git checkout pull/3874`

Update a local copy of the PR: \
`$ git checkout pull/3874` \
`$ git pull https://git.openjdk.java.net/jdk pull/3874/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3874`

View PR using the GUI difftool: \
`$ git pr show -t 3874`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3874.diff">https://git.openjdk.java.net/jdk/pull/3874.diff</a>

</details>

simonis commented on test/hotspot/gtest/logging/test_logDecorations.cpp:
```diff
@@ -32,12 +32,13 @@ static const LogTagSet& tagset = LogTagSetMapping<LOG_TAGS(logging, safepoint)>:
 static const LogDecorators default_decorators;
 
 TEST_VM(LogDecorations, level) {
+  char buf[1024];
```
Minor nit: in `LogFileStreamOutput::write_decorations()` you are using a temporary buffer of 255 charaters so you should use the same buffer size for testing. Ideally you can make that a constant or macro which can be reused here in the tests.

simonis commented on src/hotspot/share/logging/logDecorations.cpp:
You have no error handling here. `os::iso8601_time()` can return `NULL` and `print_raw()` will call `strlen(NULL)` which will crash. Do something like:
```
char *result = os::iso8601_time(_millis, buf, sizeof(buf), false);
st->print_raw(result == NULL ? "-1" : result);
```

Notice that the previous implementation was buggy as well. If `os::iso8601_time()` returned `NULL` it would have called `ASSERT_AND_RETURN(-1, pos)` which would have overwritten the last character in `pos` for non-product builds. It could also lead in a buffer-underflow if `pos` pointed to the very first character of an array.

tstuefe replied:
> You have no error handling here. `os::iso8601_time()` can return `NULL` and `print_raw()` will call `strlen(NULL)` which will crash. Do something like:
> 
> ```
> char *result = os::iso8601_time(_millis, buf, sizeof(buf), false);
> st->print_raw(result == NULL ? "-1" : result);
> ```

Good catch, I missed that. Note that I think os::iso8601_time() could be made nicer - in my opinion, any API returning the output buffer for convenience should not return NULL, because that defeats the purpose.

> 
> Notice that the previous implementation was buggy as well. If `os::iso8601_time()` returned `NULL` it would have called `ASSERT_AND_RETURN(-1, pos)` which would have overwritten the last character in `pos` for non-product builds. It could also lead in a buffer-underflow if `pos` pointed to the very first character of an array.

You are right. Nice to see this fixed.

simonis commented on src/hotspot/share/logging/logDecorations.cpp:
Same as for `LogDecorations::print_time_decoration()`.

tstuefe replied:
+1

simonis commented on src/hotspot/share/logging/logDecorations.cpp:
Yes, that would be desirable and probably not too hard :)

Otherwise you'll might have an uninitialized `tmp` array if the first tag doesn't fir into the array and  `LogTagSet::label()` returns `-1`. I understand that that's more a theoretical problem because at least one tag should always fir within 255 characters :)

tstuefe replied:
My fist version had this but Xin mentioned the patch was too large :) I'll reinstate my first solution.

simonis reviewed (changes requested):
Hi Thomas,

your change looks good and I like it :)

Am I right that you've changed the 256-char-for-all-decorators limit to a 255-char limit for a single decorator (which is obviously much better)?

Find my other comments inline.

Thank you and best regards,
Volker

YaSuenag commented on src/java.desktop/share/native/libharfbuzz/hb-ot-shape-complex-use-machine.hh:
```diff
@@ -375,7 +375,7 @@ hb_iter_with_fallback_t<machine_index_t<Iter>,
 {
         machine_index_t (const Iter& it) : it (it) {}
-        machine_index_t (const machine_index_t& o) : it (o.it) {}
+        machine_index_t (const machine_index_t& o) : hb_iter_with_fallback_t<machine_index_t<Iter>, typename Iter::item_t>(o), it (o.it) {}
```
This change does not seem to be for this PR...

tstuefe replied:
Yes, this is a harfbuzz related build fix, see initial note. I'll remove it.

tstuefe commented on src/hotspot/share/logging/logDecorations.hpp:
```diff
@@ -27,21 +27,26 @@
+  const jlong _millis;            // for "time", "utctime", "timemillis"
+  const jlong _nanos;             // for "timenanos"
+  const double _elapsed_seconds;  // for "uptime", "uptimemillis", "uptimenanos"
+  const int _pid;                 // for "pid"
```
Good point. Might even be more because of padding.

tstuefe replied:
... hah, wouldn't you know, its down to 48 bytes :-)

BTW I think if we look closer we can probably squeeze this even more.

E.g. _millis and _nanos are kind of the same thing; for _elapsedSeconds a float may be sufficient instead of a double; and tagset could be made a bitmaks (it really really should be one) and passed as 32bit value object instead of as a reference. But I wanted to keep the patch small and concise for now.

tstuefe commented:
> Hi Thomas,
> 
> your change looks good and I like it :)

:-)

> 
> Am I right that you've changed the 256-char-for-all-decorators limit to a 255-char limit for a single decorator (which is obviously much better)?

yes - and even that can be improved by using a fileStream directly, not raw jio_fprintf.

> 
> Find my other comments inline.
> 
> Thank you and best regards,
> Volker

Thanks!

tstuefe commented:
Hi @simonis , @YaSuenag , thanks for your Reviews!

New version - mostly driven by Volkers feedback:

- I now cache the pid, which further reduces the size of this object to 48. As mentioned in https://github.com/openjdk/jdk/pull/3874#discussion_r627326425, one could further reduce it but I think this is already fine.
- I added `LogDecorations::max_decoration_size` and used it for temp buffers. Note that this limit would be easy to remove by using `fileStream` in `LogFileStreamOutput::write_decorations()` - however, `fileStream` misses some functionality and I wanted to limit the patch size.
- I fixed the error handling for `os::iso8601_time()`
- I added a overload to `LogTagSet` which takes an `outputStream` - this is more elegant and prevents the copy. I also added a gtest for this function
- I removed the harfbuzz build fix

Thanks for reviewing!

..Thomas

YaSuenag commented on src/hotspot/share/logging/logDecorations.cpp:
nit: "29" might be better to define some constant e.g. "ISO8601_LEN". But it should might be better to happen in #3869.

tstuefe replied:
> Looks good!

Thanks! I'll add the constant in the other PR

YaSuenag reviewed (approved):
Looks good!

tstuefe commented on src/hotspot/share/logging/logDecorations.cpp:
```diff
@@ -25,117 +25,114 @@
-        host_name = old_value;
-      }
+      // Use raw malloc here
+      return ::strdup(buffer);
```
Hi Xin, I think you review an old version, I reverted that coding.

But the point about using a pointer and resolving hostname lazily on first access is to prevent paying anything if logging is off or logging is on but the hostname decorator is not used. For a static char array you always have to pay.

navyxliu replied:
Sorry. forget it.   I commented yesterday and just sent out. 
I see you revert host_name(). The reason to use lazy initialization is fair enough.

navyxliu commented on src/hotspot/share/logging/logTagSet.hpp:
```diff
@@ -110,6 +112,7 @@ class LogTagSet {
   void update_decorators(const LogDecorators& decorator = LogDecorators::None);
 
+  void label(outputStream* st, const char* separator = ",") const;
   int label(char *buf, size_t len, const char* separator = ",") const;
```
This old label() is only used in LogTagset. You can make it private.

Further,  no one consumes its return value now.  you can change its rettypea void.  it will avoid different return types for overloaded member functions.

tstuefe replied:
I thought so too, but the gtests still test this interface. Seeing that its not used in the hs codebase anymore I could remove it, but would have to rewrite all gtests to use the new interface instead. I'd rather avoid this to keep the patch small. Lets do this in a different RFE.

navyxliu replied:
understood. fair enough.

navyxliu commented on src/hotspot/share/logging/logDecorations.cpp:
```diff
@@ -53,89 +49,91 @@ const char* LogDecorations::host_name() {
+{
+}
+
+void LogDecorations::print_decoration(LogDecorators::Decorator decorator, outputStream* st) const {
```
Is that a good idea to do a sanity check here in debug build? 

In my understanding, calling `print_#name##decoration()` for a decorator which was not set in ctor's decorators is undefined.  Currently, your code is safe because you carefully initialize each member variable in initializer list. You can't assume follower programmers do the same. 

If you save the variable `decorators`(a bitmask), you can do sanity check here.  This can be done in debug build. eg. 
`assert(_decorators.is_decorator(decorator), "sanity check");`

tstuefe replied:
Hmm, would not the compiler warn you if you leave member initialization out?

Your idea is not bad, but I usually avoid adding debug only members because you end up with different memory size and layout of data structures between debug and opt builds, which may have strange consequences when hunting bugs.

Idk. I can do this here, I have no strong feelings either way.

navyxliu replied:
C++ won't warn you if you don't initialize member variables. Unlike Java,  C++ even doesn't zero them for you by default.  

eg. 
```
#include <stdio.h>
class T{
public:
    int a, b, c;
    T() : b(0) {}
};

int main() {
    T t;
    printf("%d\n", t.a); // random value on stack
    printf("%d\n", t.b); // 0
    printf("%d\n", t.c); // random value on stack
}
```

It's relatively safe now because all member variables you added to LogDecorations are "scalars".  If we extend an object decorator, printing an uninitialized object may result in undefined behavior.   

I am fine with/without it. can leave it to the future.

tstuefe replied:
Okay, I'll add the check

tstuefe replied:
(I think static checkers like Sonarcloud would pick those simple errors up)

tstuefe commented:
@navyxliu @simonis @YaSuenag 
Thanks for your feedbacks!

Latest iteration:
- changed help text printing to use the new `LogTagSet::label(outputStream)` interface
- added Xin's proposal of a debug check testing if someone prints an uninitialized decorator value.

Thanks, Thomas

YaSuenag commented:
I just approved #3869, so I hope to use `os::iso8601_timestamp_size` in this PR.

navyxliu commented:
I guess you will revert code change of JDK-8266536. 
For other parts in this patch, LGTM.

navyxliu commented:
I guess you will revert code change of JDK-8266536. 
For other parts in this patch, LGTM.

tstuefe commented:
> I guess you will revert code change of JDK-8266536.
Sure

> For other parts in this patch, LGTM.
Thanks!

tstuefe commented:
@simonis : if you approve, I will push.

All nightlies at SAP ran through fine.

..thomas

tstuefe commented:
> Hi Thomas,
> 
> everything looks fine now except the comments below (i.e. didn't you wanted to use the new `iso8601_timestamp_size`?).
> 
> If that 's clarified I'll approve this PR.
> 
> Thanks,
> Volker

Hi Volker,

thanks! Yes, I missed that, will change it before pushing.

Cheers, Thomas

tstuefe commented:
> Hi Thomas,
> 
> everything looks fine now except the comments below (i.e. didn't you wanted to use the new `iso8601_timestamp_size`?).
> 
> If that 's clarified I'll approve this PR.
> 
> Thanks,
> Volker

@simonis  done. You can approve.

Has been nice to have you as a Reviewer again ;-)

simonis reviewed (approved):
It has been a pleasure :)

Approved.
